include/cuco/detail/static_map/static_map_ref.inl

```diff
@@ -549,6 +551,225 @@ class operator_impl<
+  using ref_type = static_map_ref<Key, T, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>;
+  using key_type = typename base_type::key_type;
+  using value_type = typename base_type::value_type;
+
+  static constexpr auto cg_size     = base_type::cg_size;
+  static constexpr auto window_size = base_type::window_size;
+
+  static_assert(sizeof(T) == 4 or sizeof(T) == 8,
+                "sizeof(mapped_type) must be either 4 bytes or 8 bytes.");
+
+ public:
+  /**
+   * @brief Inserts a key-value pair `{k, v}` if it's not present in the map. Otherwise, applies
+   `Op`
+   * binary function to the mapped_type corresponding to the key `k` and the value `v`.
+   *
+   * @tparam Value Input type which is implicitly convertible to 'value_type'
+   * @tparam Op Callable type which is used as apply operation and can be
+   *   called with arguments as Op(cuda::atomic_ref<T, Scope>, T). Op strictly must
+   *   have this signature to atomically apply the operation.
+
+   * @param value The element to insert
+   * @param op The callable object to perform binary operation between existing value at the slot
+   *  and the element to insert.
+   */
+
+  template <typename Value, typename Op>
+  __device__ void insert_or_apply(Value const& value, Op op) noexcept
+  {
+    static_assert(cg_size == 1, "Non-CG operation is incompatible with the current probing scheme");
+
+    static_assert(
+      std::is_invocable_v<Op, cuda::atomic_ref<T, Scope>, T>,
+      "insert_or_apply expects `Op` to be a callable as `Op(cuda::atomic_ref<T, Scope>, T)`");
+
+    ref_type& ref_ = static_cast<ref_type&>(*this);
+
+    auto const val       = ref_.impl_.heterogeneous_value(value);
+    auto const key       = ref_.impl_.extract_key(val);
+    auto& probing_scheme = ref_.impl_.probing_scheme();
+    auto storage_ref     = ref_.impl_.storage_ref();
+    auto probing_iter    = probing_scheme(key, storage_ref.window_extent());
+
+    while (true) {
+      auto const window_slots = storage_ref[*probing_iter];
+
+      for (auto& slot_content : window_slots) {
+        auto const eq_res =
+          ref_.impl_.predicate_.operator()<is_insert::YES>(key, slot_content.first);
+
+        // If the key is already in the container, update the payload and return
+        if (eq_res == detail::equal_result::EQUAL) {
+          auto const intra_window_index = thrust::distance(window_slots.begin(), &slot_content);
+          op(
+            cuda::atomic_ref<T, Scope>{
+              ((storage_ref.data() + *probing_iter)->data() + intra_window_index)->second},
+            val.second);
+          return;
+        }
+        if (eq_res == detail::equal_result::AVAILABLE) {
+          auto const intra_window_index = thrust::distance(window_slots.begin(), &slot_content);
+          if (attempt_insert_or_apply(
+                (storage_ref.data() + *probing_iter)->data() + intra_window_index, value, op)) {
+            return;
```

Comment:
Uhhh I totally forgot about this idea.

Currently we're copying the first value into the slot and only subsequent values that land on it will call `apply`. Let's say we're doing a sum reduction with values `{3, 4, 5}` associated to the same key. First the `3` will be copied into the slot and then the next occurrence will call `3+4=7` and the next one `7+5=12` which is the correct result.
If we `apply` the initial value after the first insert then the result would be `3 +3 +4+5=15` which is incorrect.

But here comes the catch:
If we can ensure that the `empty_value_sentinel` is equal to the neutral element of the reduction operator (in this example `0` for a sum reduction), then we don't have to copy the value first but can just call `apply` for the first occurrence. With the above example this leads to this sequence: `0+3+4+5=12` which is still correct. Since `fetch_add` should be faster than a CAS this potentially would bring a speedup.

There is one tricky part about the design here: The user needs to inform us what the neutral element is (maybe as an optional argument that would enable this code path in case `empty_value_sentinel == neutral_element`).

Reply:
> If we apply the initial value after the first insert then the result would be 3 +3 +4+5=15 which is incorrect.

Ah OK.

The neutral payload idea seems too error-prone. We can stick with the current `attempt_insert_or_apply` logic for now.